app/controllers/api/avatars.php

```
@@ -49,11 +53,92 @@
+
+                $verificationId = $oauth2->getUserID($accessToken);
+
+                if (empty($verificationId)) {
+                    throw new \Exception("Locked tokens."); // Race codition, handeled in catch
+                }
+
+                $session
+                    ->setAttribute('providerAccessToken', $accessToken)
+                    ->setAttribute('providerRefreshToken', $refreshToken)
+                    ->setAttribute('providerAccessTokenExpiry', DateTime::addSeconds(new \DateTime(), (int)$oauth2->getAccessTokenExpiry('')));
+
+                Authorization::skip(fn () => $dbForProject->updateDocument('sessions', $session->getId(), $session));
+
+                $dbForProject->deleteCachedDocument('users', $user->getId());
+            } catch (Throwable $err) {
+                $index = 0;
+                do {
+                    $previousAccessToken = $session->getAttribute('providerAccessToken');
+
+                    $user = Authorization::skip(fn () => $dbForConsole->getDocument('users', $userId));
+                    $sessions = $user->getAttribute('sessions', []);
+                    $session = $sessions[0] ?? new Document();
+                    $accessToken = $session->getAttribute('providerAccessToken');
+
+                    if ($accessToken !== $previousAccessToken) {
+                        break;
+                    }
+
+                    $index++;
+                    \usleep(500000);
+                } while ($index < 10);
+            }
+        }
+
+        $oauth2 = new $className($appId, $appSecret, '', [], []);
+        $githubUser = $oauth2->getUserSlug($accessToken);
+        $githubId = $oauth2->getUserID($accessToken);
+
+        return [
+            'name' => $githubUser,
+            'id' => $githubId
+        ];
+    } catch (Exception $err) {
+        \var_dump($err->getMessage());
+        \var_dump($err->getTraceAsString());
+        \var_dump($err->getLine());
+        \var_dump($err->getFile());
```

we should rethrow the exception so it can be handled by our normal exception handling flow